typescript/alert-center/src/utilities.ts

```diff
@@ -0,0 +1,165 @@
+        a.alert_status = AlertStatus.ASSIGNED;
+      }
+    }
+    for (const p of employees) {
+      if (p.expiration) {
+        p.timeLeft = Math.round((p.expiration.getTime() - new Date().getTime())/1000);
+      }
+    }
+    return {alerts, employees};
+  }  
+
+  @Transaction()
+  static async cleanStaff(ctx: KnexTransactionContext) {
+    await ctx.client<Employee>('employee').whereNull('alert_id').delete();
+  }
+
+  @Transaction()
+  static async cleanAlerts(ctx: KnexTransactionContext) {
+    await ctx.client<AlertEmployee>('alert_employee').where({alert_status: AlertStatus.RESOLVED}).delete();
+  }
+
+  @Transaction()
+  static async getMaxId(ctx: KnexTransactionContext) {
+    const result = await ctx.client<AlertEmployee>('alert_employee').max('alert_id', { as: 'mid' }).first();
+    if (result ) {
+      return result.mid;
+    }
+    return -1;
+  }
+
+  @Transaction()
+  static async addAlert(ctx: KnexTransactionContext, message: AlertWithMessage) {
+    await ctx.client<AlertEmployee>('alert_employee').insert({
+      alert_id: message.alert_id,
+      alert_status: message.alert_status,
+      message: message.message,
+      employee_name: null,
+    }).onConflict(['alert_id']).ignore();
+  }
+
+  @Transaction()
+  static async getUserAssignment(ctx: KnexTransactionContext, employee_name: string, expiration: Date, @ArgOptional more_time: boolean | undefined) {
+    let employees = await ctx.client<Employee>('employee').where({employee_name}).select();
+    let newAssignment = false;
+    if (!employees.length) {
+      await ctx.client<Employee>('employee').insert({employee_name, alert_id: null, expiration: null});
+      employees = await ctx.client<Employee>('employee').where({employee_name}).select();
+    }
+    let alert : AlertEmployee[] = [];
+    if (employees[0].alert_id && more_time) {
+      // Extend time
+      ctx.logger.info(`Extending time for ${employee_name} on ${employees[0].alert_id}`);
+      if (employees[0].expiration?.getTime() ?? 0 < expiration.getTime()) {
+        employees[0].expiration = expiration;
+        await ctx.client<Employee>('employee').where({employee_name}).update({expiration});
+      }
+    } 
+    else if (employees[0].alert_id) {
+      alert = await ctx.client<AlertEmployee>('alert_employee').where({alert_id: employees[0].alert_id}).select();
+      return {employee: employees[0], newAssignment, alert};
+    }
+    else {
+      // Try to find assignment
+      const op = await ctx.client<AlertEmployee>('alert_employee').whereNull('employee_name').orderBy(['alert_id']).first();
+      if (op) {
+        op.employee_name = employee_name;
+        const alert_id = op.alert_id;
+        employees[0].alert_id = op.alert_id;
+        employees[0].expiration = expiration;
+        await ctx.client<Employee>('employee').where({employee_name}).update({alert_id, expiration});
+        await ctx.client<AlertEmployee>('alert_employee').where({alert_id}).update({employee_name});
```

Comment:
Can we use a "returning clause" here? Then we don't need to select the alert_employee table again at the end.

Reply:
This is the piece I rewrote to make "more_time" easier to read. See if you like the new version?

Reply:
Looks good!